In [1]:
file_path = r"C:\unibo-dtm-ml-2526-cervical-cancer-predictor\data\raw.csv"
with open(file_path, 'r') as f:
    lines = f.readlines()
print('Read {} lines'.format(len(lines)))

Read 859 lines


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df = pd.read_csv(file_path)
df

Age Number of sexual partners First sexual intercourse  \
0     18                       4.0                     15.0   
1     15                       1.0                     14.0   
2     34                       1.0                        ?   
3     52                       5.0                     16.0   
4     46                       3.0                     21.0   
..   ...                       ...                      ...   
853   34                       3.0                     18.0   
854   32                       2.0                     19.0   
855   25                       2.0                     17.0   
856   33                       2.0                     24.0   
857   29                       2.0                     20.0   

    Num of pregnancies Smokes Smokes (years) Smokes (packs/year)  \
0                  1.0    0.0            0.0                 0.0   
1                  1.0    0.0            0.0                 0.0   
2                  1.0    0.0            0.0                 0.0   
3                  4.0    1.0           37.0                37.0   
4                  4.0    0.0            0.0                 0.0   
..                 ...    ...            ...                 ...   
853                0.0    0.0            0.0                 0.0   
854                1.0    0.0            0.0                 0.0   
855                0.0    0.0            0.0                 0.0   
856                2.0    0.0            0.0                 0.0   
857                1.0    0.0            0.0                 0.0   

    Hormonal Contraceptives Hormonal Contraceptives (years)  IUD  ...  \
0                       0.0                             0.0  0.0  ...   
1                       0.0                             0.0  0.0  ...   
2                       0.0                             0.0  0.0  ...   
3                       1.0                             3.0  0.0  ...   
4                       1.0                            15.0  0.0  ...   
..                      ...                             ...  ...  ...   
853                     0.0                             0.0  0.0  ...   
854                     1.0                             8.0  0.0  ...   
855                     1.0                            0.08  0.0  ...   
856                     1.0                            0.08  0.0  ...   
857                     1.0                             0.5  0.0  ...   

    STDs: Time since first diagnosis STDs: Time since last diagnosis  \
0                                  ?                               ?   
1                                  ?                               ?   
2                                  ?                               ?   
3                                  ?                               ?   
4                                  ?                               ?   
..                               ...                             ...   
853                                ?                               ?   
854                                ?                               ?   
855                                ?                               ?   
856                                ?                               ?   
857                                ?                               ?   

    Dx:Cancer Dx:CIN Dx:HPV Dx Hinselmann Schiller Citology Biopsy  
0           0      0      0  0          0        0        0      0  
1           0      0      0  0          0        0        0      0  
2           0      0      0  0          0        0        0      0  
3           1      0      1  0          0        0        0      0  
4           0      0      0  0          0        0        0      0  
..        ...    ...    ... ..        ...      ...      ...    ...  
853         0      0      0  0          0        0        0      0  
854         0      0      0  0          0        0        0      0  
855         0      0      0  0          0        0        1      0  

In [8]:
df.columns

Index(['Age', 'Number of sexual partners', 'First sexual intercourse',
       'Num of pregnancies', 'Smokes', 'Smokes (years)', 'Smokes (packs/year)',
       'Hormonal Contraceptives', 'Hormonal Contraceptives (years)', 'IUD',
       'IUD (years)', 'STDs', 'STDs (number)', 'STDs:condylomatosis',
       'STDs:cervical condylomatosis', 'STDs:vaginal condylomatosis',
       'STDs:vulvo-perineal condylomatosis', 'STDs:syphilis',
       'STDs:pelvic inflammatory disease', 'STDs:genital herpes',
       'STDs:molluscum contagiosum', 'STDs:AIDS', 'STDs:HIV',
       'STDs:Hepatitis B', 'STDs:HPV', 'STDs: Number of diagnosis',
       'STDs: Time since first diagnosis', 'STDs: Time since last diagnosis',
       'Dx:Cancer', 'Dx:CIN', 'Dx:HPV', 'Dx', 'Hinselmann', 'Schiller',
       'Citology', 'Biopsy'],
      dtype='object')

### Dataset Description

1. **`Age`**: Patient's age in years.
2. **`Number of sexual partners`**: Lifetime total of sexual partners.
3. **`First sexual intercourse`**: Age (years) at first sexual encounter.
4. **`Num of pregnancies`**: Total number of pregnancies.
5. **`Smokes`**: Binary indicator (1=Yes, 0=No) if the patient currently smokes.
6. **`Smokes (years)`**: Total duration of smoking habit in years.
7. **`Smokes (packs/year)`**: Average number of cigarette packs consumed per year.
8. **`Hormonal Contraceptives`**: Binary indicator (1=Yes, 0=No) for hormonal birth control use.
9. **`Hormonal Contraceptives (years)`**: Total duration of hormonal contraceptive use in years.
10. **`IUD`**: Binary indicator (1=Yes, 0=No) for the presence of an Intrauterine Device.
11. **`IUD (years)`**: Total duration of IUD use in years.
12. **`STDs`**: Binary indicator (1=Yes, 0=No) for any history of sexually transmitted diseases.
13. **`STDs (number)`**: Total count of different types of STDs contracted by the patient.
14. **`STDs:condylomatosis`**: Binary indicator for history of genital warts.
15. **`STDs:cervical condylomatosis`**: Binary indicator for warts specifically on the cervix.
16. **`STDs:vaginal condylomatosis`**: Binary indicator for warts specifically in the vagina.
17. **`STDs:vulvo-perineal condylomatosis`**: Binary indicator for warts on external genitalia.
18. **`STDs:syphilis`**: Binary indicator for history of Syphilis infection.
19. **`STDs:pelvic inflammatory disease`**: Binary indicator for history of PID.
20. **`STDs:genital herpes`**: Binary indicator for history of Herpes infection.
21. **`STDs:molluscum contagiosum`**: Binary indicator for history of Molluscum contagiosum.
22. **`STDs:AIDS`**: Binary indicator for history of AIDS (typically 0 in this dataset).
23. **`STDs:HIV`**: Binary indicator for history of HIV infection.
24. **`STDs:Hepatitis B`**: Binary indicator for history of Hepatitis B.
25. **`STDs:HPV`**: Binary indicator for history of Human Papillomavirus.
26. **`STDs: Number of diagnosis`**: Total number of confirmed STD diagnoses.
27. **`STDs: Time since first diagnosis`**: Years elapsed since the patient's first STD diagnosis.
28. **`STDs: Time since last diagnosis`**: Years elapsed since the patient's most recent STD diagnosis.
29. **`Dx:Cancer`**: Clinical history indicating if the patient was previously diagnosed with cancer.
30. **`Dx:CIN`**: History of Cervical Intraepithelial Neoplasia (precancerous condition).
31. **`Dx:HPV`**: History of a clinical HPV diagnosis.
32. **`Dx`**: General indicator of any previous medical diagnosis.
33. **`Hinselmann`**: Target 1: Result of colposcopy (abnormal visual exam).
34. **`Schiller`**: Target 2: Result of Lugol's iodine staining test (staining abnormalities).
35. **`Citology`**: Target 3: Result of a Pap smear (cytological cell analysis).
36. **`Biopsy`**: Target 4: Result of tissue sample analysis (histological gold standard).



### Data Profiling

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Age                                 858 non-null    int64 
 1   Number of sexual partners           858 non-null    object
 2   First sexual intercourse            858 non-null    object
 3   Num of pregnancies                  858 non-null    object
 4   Smokes                              858 non-null    object
 5   Smokes (years)                      858 non-null    object
 6   Smokes (packs/year)                 858 non-null    object
 7   Hormonal Contraceptives             858 non-null    object
 8   Hormonal Contraceptives (years)     858 non-null    object
 9   IUD                                 858 non-null    object
 10  IUD (years)                         858 non-null    object
 11  STDs                                858 non-null    object

In [27]:
df.describe(include='all')

Age Number of sexual partners First sexual intercourse  \
count   858.000000                       858                      858   
unique         NaN                        13                       22   
top            NaN                       2.0                     15.0   
freq           NaN                       272                      163   
mean     26.820513                       NaN                      NaN   
std       8.497948                       NaN                      NaN   
min      13.000000                       NaN                      NaN   
25%      20.000000                       NaN                      NaN   
50%      25.000000                       NaN                      NaN   
75%      32.000000                       NaN                      NaN   
max      84.000000                       NaN                      NaN   

       Num of pregnancies Smokes Smokes (years) Smokes (packs/year)  \
count                 858    858            858                 858   
unique                 12      3             31                  63   
top                   1.0    0.0            0.0                 0.0   
freq                  270    722            722                 722   
mean                  NaN    NaN            NaN                 NaN   
std                   NaN    NaN            NaN                 NaN   
min                   NaN    NaN            NaN                 NaN   
25%                   NaN    NaN            NaN                 NaN   
50%                   NaN    NaN            NaN                 NaN   
75%                   NaN    NaN            NaN                 NaN   
max                   NaN    NaN            NaN                 NaN   

       Hormonal Contraceptives Hormonal Contraceptives (years)  IUD  ...  \
count                      858                             858  858  ...   
unique                       3                              41    3  ...   
top                        1.0                             0.0  0.0  ...   
freq                       481                             269  658  ...   
mean                       NaN                             NaN  NaN  ...   
std                        NaN                             NaN  NaN  ...   
min                        NaN                             NaN  NaN  ...   
25%                        NaN                             NaN  NaN  ...   
50%                        NaN                             NaN  NaN  ...   
75%                        NaN                             NaN  NaN  ...   
max                        NaN                             NaN  NaN  ...   

       STDs: Time since first diagnosis STDs: Time since last diagnosis  \
count                               858                             858   
unique                               19                              19   
top                                   ?                               ?   
freq                                787                             787   
mean                                NaN                             NaN   
std                                 NaN                             NaN   
min                                 NaN                             NaN   
25%                                 NaN                             NaN   
50%                                 NaN                             NaN   
75%                                 NaN                             NaN   
max                                 NaN                             NaN   

         Dx:Cancer      Dx:CIN      Dx:HPV          Dx  Hinselmann  \
count   858.000000  858.000000  858.000000  858.000000  858.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean      0.020979    0.010490    0.020979    0.027972    0.040793   
std       0.143398    0.101939    0.143398    0.164

#### Downsides list 
As things stand, we are definetely not happy with what we've understood so far: 
- Missing values are not simply empty cells, but rather they're marked by a "?" sign implying that the value is unkown. This means that there are no missing entries (count=858 for all columns), yet those cells that would have been empty are filled with the aforementioned string value. 
- Columns containing such missing values are characterized by the "object" data type because they both store numerical (normally registered entries) and string information (the "?" marks indeed). 
- No mathematically-backed formulae can be performed on object values, as we see from the table above, thus meaning that no relevant distribution information can be inferred by columns affected by such values. 
- I presume that all the other columns with DType == int64 are complete, but this will be certified upon deeper inspection of the dataset. 



### Room for work
1. Deal with the missing values. First of all, I need to get rid of the "?" signs. Further preprocessing steps will be covered on those cells affected by the removal. 


In [ ]:
# Reshaping the dataframe. This time the '?' values are considered as NaN values.
df = pd.read_csv(r'C:\unibo-dtm-ml-2526-cervical-cancer-predictor\data\raw.csv',na_values='?')
df

Age  Number of sexual partners  First sexual intercourse  \
0     18                        4.0                      15.0   
1     15                        1.0                      14.0   
2     34                        1.0                       NaN   
3     52                        5.0                      16.0   
4     46                        3.0                      21.0   
..   ...                        ...                       ...   
853   34                        3.0                      18.0   
854   32                        2.0                      19.0   
855   25                        2.0                      17.0   
856   33                        2.0                      24.0   
857   29                        2.0                      20.0   

     Num of pregnancies  Smokes  Smokes (years)  Smokes (packs/year)  \
0                   1.0     0.0             0.0                  0.0   
1                   1.0     0.0             0.0                  0.0   
2                   1.0     0.0             0.0                  0.0   
3                   4.0     1.0            37.0                 37.0   
4                   4.0     0.0             0.0                  0.0   
..                  ...     ...             ...                  ...   
853                 0.0     0.0             0.0                  0.0   
854                 1.0     0.0             0.0                  0.0   
855                 0.0     0.0             0.0                  0.0   
856                 2.0     0.0             0.0                  0.0   
857                 1.0     0.0             0.0                  0.0   

     Hormonal Contraceptives  Hormonal Contraceptives (years)  IUD  ...  \
0                        0.0                             0.00  0.0  ...   
1                        0.0                             0.00  0.0  ...   
2                        0.0                             0.00  0.0  ...   
3                        1.0                             3.00  0.0  ...   
4                        1.0                            15.00  0.0  ...   
..                       ...                              ...  ...  ...   
853                      0.0                             0.00  0.0  ...   
854                      1.0                             8.00  0.0  ...   
855                      1.0                             0.08  0.0  ...   
856                      1.0                             0.08  0.0  ...   
857                      1.0                             0.50  0.0  ...   

     STDs: Time since first diagnosis  STDs: Time since last diagnosis  \
0                                 NaN                              NaN   
1                                 NaN                              NaN   
2                                 NaN                              NaN   
3                                 NaN                              NaN   
4                                 NaN                              NaN   
..                                ...                              ...   
853                               NaN                              NaN   
854                               NaN                              NaN   
855                               NaN                              NaN   
856                               NaN                              NaN   
857                               NaN                              NaN   

     Dx:Cancer  Dx:CIN  Dx:HPV  Dx  Hinselmann  Schiller  Citology  Biopsy  
0            0       0       0   0           0         0         0       0  
1            0       0       0   0           0         0         0       0  
2            0       0       0   0           0         0         0       0  
3            1       0       1   0           0         0         0       0  
4            0       0       0   0           0         0         0       0  
..         ...     ...     ...  ..         ...       ...       ...     ...  
853          0       0       0

In [ ]:
df.infer_objects().dtypes



Age                                     int64
Number of sexual partners             float64
First sexual intercourse              float64
Num of pregnancies                    float64
Smokes                                float64
Smokes (years)                        float64
Smokes (packs/year)                   float64
Hormonal Contraceptives               float64
Hormonal Contraceptives (years)       float64
IUD                                   float64
IUD (years)                           float64
STDs                                  float64
STDs (number)                         float64
STDs:condylomatosis                   float64
STDs:cervical condylomatosis          float64
STDs:vaginal condylomatosis           float64
STDs:vulvo-perineal condylomatosis    float64
STDs:syphilis                         float64
STDs:pelvic inflammatory disease      float64
STDs:genital herpes                   float64
STDs:molluscum contagiosum            float64
STDs:AIDS                         

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 858 entries, 0 to 857
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Age                                 858 non-null    int64  
 1   Number of sexual partners           832 non-null    float64
 2   First sexual intercourse            851 non-null    float64
 3   Num of pregnancies                  802 non-null    float64
 4   Smokes                              845 non-null    float64
 5   Smokes (years)                      845 non-null    float64
 6   Smokes (packs/year)                 845 non-null    float64
 7   Hormonal Contraceptives             750 non-null    float64
 8   Hormonal Contraceptives (years)     750 non-null    float64
 9   IUD                                 741 non-null    float64
 10  IUD (years)                         741 non-null    float64
 11  STDs                                753 non-n

In [38]:
df['Smokes'].unique()

array([ 0.,  1., nan])